[WIP] [Docs] Docs deps upgrade

peytondmurray opened the pull request:
## WIP PR; here for visibility

## Why are these changes needed?

This PR updates the documentation build dependencies to the latest versions. Changes to the docs needed to accommodate this upgrade:

- Remove `margin` directives; they are only implemented in `sphinx-book-theme`, which we are no longer using. These have all been converted to `note` directives.
- `sphinx-external-toc` has been removed. There are a lot of orphaned pages; I've added these to the appropriate pages.
- Removed `LandingPageBG.jpg`: the docs landing page intended to use this but has accidentally been targeting `LandingPageBG.JPG`, and therefore has been failing to fetch the image. I'm just removing it for now, as it doesn't adapt to dark mode anyway.
- Removed a reference to `cluster/kubernetes/troubleshooting/troubleshooting.html#gpu-multitenancy`, there is no `gpu-multitenancy` ref. There are a number of references in notebooks that have been converted to explicit myst references, fixing some issues with reference ambiguity.

## Development changes to be addressed before draft is ready for review

- [x] `Makefile` changes to speed up development
- [x] Reenable `fail_on_warning` in `.readthedocs.yml`
- [x] Search for all pngs, investigate to make sure dark-compatible. If incompatible, replace with SVG **Update: Partially complete; have finished most critical ones, will defer on rest.**
- [x] CSAT widget
- [ ] Anyscale logo is kind of hard to see on dark background - find another solution **Deferred**
- [x] Ask AI widget
- [x] Example gallery sidebar and searchbar
- [x] Override pydata-sphinx-theme colors to use colors more in line with Ray branding

## Related issue number

Closes https://github.com/ray-project/ray/issues/37944.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

can-anyscale commented on doc/Makefile:
```diff
@@ -2,7 +2,7 @@
 
 # You can set these variables from the command line.
-SPHINXOPTS    = -a -E
+SPHINXOPTS    = -W
```
dummy question, how does this change the build process?

peytondmurray replied:
I'll change this back before merging this, but:

- `-a -E` makes sphinx rebuild the docs each time instead of using the existing build artifacts
- `-W` makes sphinx turn any build warnings into build _errors_, stopping the build.

can-anyscale commented on doc/requirements-doc.txt:
```diff
@@ -1,35 +1,34 @@
 
 git+https://github.com/ray-project/tune-sklearn@master#tune-sklearn
-watchfiles # Required because sphinx-click doesn't support mocking
+watchfiles==0.20.0 # Required because sphinx-click doesn't support mocking
```
look like there are some issues with resolving conflicts: https://buildkite.com/ray-project/premerge/builds/8084#018b1ad2-e899-4517-9744-0932f9dacbca/198-364, to update them all to the latest version, I would suggest to unpin all versions and let the ci system resolve the versions itself

The ci system will resolve all conflict and output the content in this job: https://buildkite.com/ray-project/premerge/builds/8084#018b1ad2-e899-4517-9744-0932f9dacbca/198-364, which you can copy the content into this file https://github.com/ray-project/ray/blob/master/python/requirements_compiled.txt

Let me know if these doesn't make sense. Thankks

peytondmurray replied:
So I just tested unpinning these and constraining them with `requirements_compiled.txt` but that doesn't work; `myst-nb` is constrained by `test-requirements.txt` :/. Same for `myst-parser`. For now I'm not going to spend any time on this until later on when we can come up with a way to reconcile these differences.

can-anyscale replied:
Ah I see, doc test requirements are declared in test-requirements.txt as well ;). Let me give this a shot and if it works I can put up a PR that you can reuse.

can-anyscale replied:
My attempt: https://github.com/ray-project/ray/pull/40345

simran-2797 commented:
Few comments from the design side: 
1. I have a design for dark mode for the home page. We can use that - https://www.figma.com/file/fRYp3ZBGbpg9dlu3WlBYY9/Quansight---Example-gallery-and-CSAT?type=design&node-id=288%3A546&mode=design&t=HZiNlsIehbqxIXH5-1
2. Looks like there are 3 different button states instead of 2 for dark and light
![Screenshot 2023-10-31 at 3 45 17 PM](https://github.com/ray-project/ray/assets/23429473/57679b87-6291-49b6-9a0c-b728c1c1ff54)
![Screenshot 2023-10-31 at 3 45 47 PM](https://github.com/ray-project/ray/assets/23429473/39d6f828-88b8-4955-b1ae-ccfdfcbb095d)
![Screenshot 2023-10-31 at 3 46 03 PM](https://github.com/ray-project/ray/assets/23429473/ca2900bf-e6e8-4e85-bf86-4882b8bde897)
3. The example gallery is not on a separate page. The filters are missing. 
![Screenshot 2023-10-31 at 3 47 08 PM](https://github.com/ray-project/ray/assets/23429473/803a3e32-4343-4407-86d3-5639f8522fc7)

peytondmurray commented:
@simran-2797 Thanks for taking a look. I'll implement the color changes you made in the mockups.

> Looks like there are 3 different button states instead of 2 for dark and light

Yes. The tooltip for the widget explains that the switch can be `dark`, `light` or `use whatever the current system theme is` modes. Is this undesirable? It's how all the docs based on `pydata-sphinx-theme` work.

Re: example gallery - this is something I still need to finish up. Thanks for the reminder!

angelinalg commented:
There is a pound (#) sign that appears at the end of the last paragraph on the RLlib index page that I don't see in master: 
<img width="724" alt="Screenshot 2023-10-31 at 4 35 58 PM" src="https://github.com/ray-project/ray/assets/122562471/5065f3fe-e749-4d8a-8806-bf92afb051ee">

angelinalg commented:
Don't forget to turn off the highlighting!

angelinalg commented:
- [ ] Replace Train top page logo with SVG
![Group 48](https://github.com/ray-project/ray/assets/122562471/c814869e-f0ba-454f-97ad-40215da91d6d)

peytondmurray commented:
> There is a pound (#) sign that appears at the end of the last paragraph on the RLlib index page that I don't see in master

That's because you're hovering over an image that was inserted as a `figure`. It gives you a way to copy a link to that image; here's the raw `rst`:

```rst
.. figure:: images/rllib-stack.svg
    :align: left
    :width: 650

    **RLlib's API stack:** Built on top of Ray, RLlib offers off-the-shelf, highly distributed
    algorithms, policies, loss functions, and default models (including the option to
    auto-wrap a neural network with an LSTM or an attention net). Furthermore, our library
    comes with a built-in Server/Client setup, allowing you to connect
    hundreds of external simulators (clients) via the network to an RLlib server process,
    which provides learning functionality and serves action queries. User customizations
    are realized via sub-classing the existing abstractions and - by overriding certain
    methods in those sub-classes - define custom behavior.
```

If this is undesirable behavior, I think there's probably ways of using other directives or turning this feature off.

angelinalg commented:
- [ ] The `Edit in GitHub` button 404's on this page: https://anyscale-ray--39766.com.readthedocs.build/en/39766/ray-observability/getting-started.html#observability-getting-started

angelinalg commented:
- [ ] `Hide Search Metrics` button at the top of the page.
<img width="863" alt="Screenshot 2023-10-31 at 5 03 49 PM" src="https://github.com/ray-project/ray/assets/122562471/9bd044c3-83b4-4711-83c5-a73f95896ebe">

angelinalg commented:
- [ ] The search box in the Example Gallery page doesn't seem to work.

angelinalg commented:
- [ ] Top nav: Change default page for `Resources` button so that it's different from the `Docs` default page.

angelinalg commented:
- [ ] Side nav: Developer Guides should be expandable.
<img width="1023" alt="Screenshot 2023-10-31 at 5 17 12 PM" src="https://github.com/ray-project/ray/assets/122562471/abd0922a-ac2e-4153-afb7-59c077ac3119">

This is what is on master: 
<img width="823" alt="Screenshot 2023-10-31 at 5 18 08 PM" src="https://github.com/ray-project/ray/assets/122562471/0ef9bf73-41b6-4f83-b974-735002f5976b">

angelinalg commented:
- [ ] SIde nav: Ray Core > User Guides should be expandable with child pages.

angelinalg commented:
- [ ] Side nav doesn't reflect the location of the current page. Is this by design?

bveeramani reviewed (commented):
Data docs overall LGTM. Nice work!

- [ ] Nit: I think we don't want the leading whitespace to be underscored?
<img width="151" alt="image" src="https://github.com/ray-project/ray/assets/26107013/5d216b74-b071-48ec-91bc-178504ffd4f1">

- [ ] Nit: This is more of a personal preference, but I think the code text is more readable without the grey background.
<img width="353" alt="image" src="https://github.com/ray-project/ray/assets/26107013/ba04e16f-e199-4c37-ac76-f84fb3e64bb3">

- [ ] When I click "User guides", I don't see the specific user guides in the sidebar. Is this expected? I feel like it's easier to read the sidebar than the list of links in the main page.

<img width="1728" alt="image" src="https://github.com/ray-project/ray/assets/26107013/7864d054-24ee-4881-980d-c17117638dda">

- [ ] Not sure if this is because we need to merge in master, or because the CSS classes are different, but looks like the tables aren't aligned 

<img width="741" alt="image" src="https://github.com/ray-project/ray/assets/26107013/59540597-6172-4408-a467-72d772854079">

peytondmurray commented:
> When I click "User guides", I don't see the specific user guides in the sidebar. Is this expected? I feel like it's easier to read the sidebar than the list of links in the main page.

@bveeramani Sidebar behavior is something that I think we haven't yet really figured out, and which isn't well sorted in the current version of the docs. On this branch we have the links defined in [sidebar.yml](https://github.com/peytondmurray/ray/blob/docs-deps-upgrade/doc/source/sidebar.yml) and that's it. Do we want _all_ possible links nested (potentially very deeply) in the sidebar? I'm happy to insert whatever links we want, and at the same time we should also think carefully about how deep nesting affects usability.

Thanks for the feedback!

bveeramani commented:
> Do we want _all_ possible links nested (potentially very deeply) in the sidebar?

No, I don't think so. I think the individual API reference pages aren't useful in the sidebar. I think the user guides are helpful, though

peytondmurray commented:
> Dark mode: Logos don't look good on this page: https://anyscale-ray--39766.com.readthedocs.build/en/39766/train/more-frameworks.html

I look the time to rewrite this: the cards weren't clickable, instead a button was being embedded inside the card leading to some awkward spacing. I added some margin to the tops of the images, this is the result:

![image](https://github.com/ray-project/ray/assets/14017872/8244d910-f290-43fd-af65-e5fb602be0dc)

This looks a little better, but as an alternative we can also replace these with SVGs that have dynamically colored text that can change according to the theme, as we do elsewhere. I think this would be better - what do you think @angelinalg?

emmyscode commented:
# Review of "Ray Clusters" section

- [ ] On the "Ray Clusters Overview" page, the "Learn Key Concepts" button/card link jumps you down past the title on the "Key Concepts" page.
- [ ] On the "Key Concepts" page, each heading (i.e. Ray Cluster, Head Node, Worker Node, Autoscaling, and Ray Jobs) when clicked shoots you up to the Ray Cluster heading.
- [ ] When I'm on a page, it's not highlighted in the sidenav, which causes me to kind of lose where I am, especially in a deeply nested section like "Ray Clusters."
- [ ] When I go to any subpage of "API References" under "Deploying VMs," the sidenav collapses completely and I lose my spot. 
- [ ] "API References" under "Deploying VMs" is section, but is not bolded or expandable.
- [ ] "Applications Guide" is a section, but is not expandable or bolded in the sidenav.
- [ ] When I go to a subpage in "Applications Guide" the sidenav collapses.
- [ ] Accessing any subpage in the "Ray Cluster Management API" causes the sidenav to collapse. Also, it's a section, but is not bolded and subitems don't appear in the dropdown.

peytondmurray commented:
- [ ] Make all blues the same

peytondmurray commented:
@emmyscode 
 
> On the "Ray Clusters Overview" page, the "Learn Key Concepts" button/card link jumps you down past the title on the "Key Concepts" page.

I'm seeing this on <docs.ray.io/en/master> as well. This is not due to any change I've made here, but the fact that the `Learn Key Concepts` button links to an anchor that is placed underneath the heading of the page. Really it should be linking to the doc, not an arbitrary sphinx ref. This is a pattern found in many places, but I think it would be best to clean this stuff up after the update is complete.


> On the "Key Concepts" page, each heading (i.e. Ray Cluster, Head Node, Worker Node, Autoscaling, and Ray Jobs) when clicked shoots you up to the Ray Cluster heading.

I'm also seeing this on <docs.ray.io/en/master>; leaving for future cleanup.

> When I'm on a page, it's not highlighted in the sidenav, which causes me to kind of lose where I am, especially in a deeply nested section like "Ray Clusters."
When I go to any subpage of "API References" under "Deploying VMs," the sidenav collapses completely and I lose my spot.
> "API References" under "Deploying VMs" is section, but is not bolded or expandable.
> "Applications Guide" is a section, but is not expandable or bolded in the sidenav.
> When I go to a subpage in "Applications Guide" the sidenav collapses.
> Accessing any subpage in the "Ray Cluster Management API" causes the sidenav to collapse. Also, it's a section, but is not bolded and subitems don't appear in the dropdown.

The rest of these are due to the fact that we don't currently have _every single_ page accessible from the sidebar - it's sort of an arbitrary collection of links at the moment. @angelinalg and I met earlier this week and decided to put every page in the sidebar until we can find a better solution, so for now that's what I'll do. I'll ping once it's ready for you to look at.

peytondmurray commented:
I think I've addressed all the feedback from the docs team. I'll split this PR into reviewable chunks against a new branch, then we'll merge that branch to `master` at which point I'll close this PR.

peytondmurray commented:
Closing now that https://github.com/ray-project/ray/pull/41115 is merged.
